Results by cookie

mkcode opened the pull request:
Hi @deski -  Following up on our discussion in #78 

This PR adds info from the peek cookie as discussed. It does not save Peek results unless the peek cookie is enabled. Always sets the cookie if peek is enabled.

In the case Peek was not enabled for a request, and is toggled on, `Peek::Views` with a `data-defer-to` are hidden and a refresh message is displayed: 

![screen shot 2015-10-12 at 8 41 31 pm](https://cloud.githubusercontent.com/assets/283496/10442611/22bb6a2e-7122-11e5-8fb3-c3fb5ed333b7.png)

dewski commented:
Thanks for opening this up! I'll take a look at this over the next few days.

dewski commented on test/controllers/requests_test.rb:
```diff
@@ -1,14 +1,30 @@
+  end
+
+  test "does not save results if the peek cookie is not set" do
+    assert_not_called(Peek.adapter, :save) do
```
There should be an assertion here that the cookie isn't set.

dewski commented on lib/peek/controller_helpers.rb:
```diff
@@ -9,6 +9,11 @@ module ControllerHelpers
 
+    def append_info_to_payload(payload)
+      super
+      payload[:peek_enabled] = request.cookies['peek'] == 'true'
```
We should move this conditional to `peek_enabled?` below and let internal method determine what the boolean is in the payload.

dewski replied:
```ruby
payload[:peek_enabled] = peek_enabled?
```

dewski commented on lib/peek/railtie.rb:
```diff
@@ -20,8 +20,11 @@ class Railtie < ::Rails::Engine
+        |_, _, _, _, payload|
+        if payload[:peek_enabled]
+          Peek.adapter.save
+          Peek.clear
```
We'll want to run clear after every request even if we don't persist any data.

dewski commented on test/controllers/requests_test.rb:
```diff
@@ -1,14 +1,30 @@
 require 'test_helper'
 
 class RequestsTest < ActionDispatch::IntegrationTest
+  include ActiveSupport::Testing::MethodCallAssertions
```
I'd rather we don't introduce mocking if possible. What I'm thinking is we should be able to do something like:

```ruby
assert_nil Peek.adapter.get(request.env['action_dispatch.request_id'])
```

To determine if the request did or didn't save any data.

dewski commented:
Thanks for being patient with me, I gave this a round of review. Let me know what you think.

mkcode commented:
Great feedback. I'll fix this up.

mkcode commented:
Hi @dewski - I updated this branch according to your feedback.

I changed the the name of the cookie test from `peek_enabled` to `peek_bar_visible`: https://github.com/peek/peek/pull/79/files#diff-5fa9351cd8e19ee995f9f3c600861393R14 in response to this comment: https://github.com/peek/peek/pull/79#discussion_r43567444. We can't reuse peek_enabled for this, because it is in a global context, while the payload flag is per request. `peek_bar_visible` is a better name for this now, but that can change if you like.

This test could be better: https://github.com/peek/peek/pull/79/files#diff-d89c7ee11ab586df930a3fb679d46386R10. It is redundant with the other tests and is actually testing if the request is saved. Ideally, this would be testing that `Peek._request_id` is set, but I could not figure out how to do this without mocking, because it is cleared after the request.

dewski commented:
> We can't reuse peek_enabled for this, because it is in a global context, while the payload flag is per request. peek_bar_visible is a better name for this now, but that can change if you like.

Could you clarify what you mentioned here? I'd like to have `peek_enabled?` be the one source for determining if the bar is to be displayed and tracked. If it's disabled it won't render the view and after the request it won't save the payload. Since both methods are in the same controller they should be able to call one another.

dewski commented on test/controllers/requests_test.rb:
```diff
@@ -4,11 +4,25 @@ class RequestsTest < ActionDispatch::IntegrationTest
+  test "the request id is set when the peek cookie is set" do
     assert_empty Peek.adapter.requests
     get '/'
     assert_not_empty Peek.adapter.requests
```
I don't think the test case changed here at all, just need to update the test body:

```ruby
assert_empty Peek.request_id
get '/'
refute_empty Peek.request_id
```

mkcode replied:
Ok! Will update this.

mkcode commented:
> Could you clarify what you mentioned here?

Sure. `peek_enabled` is meant to overridden in descendant controllers. From the readme:

```
class ApplicationController < ActionController::Base
  def peek_enabled?
    current_user.staff?
  end
end
```

If we were to reuse `peek_enabled`, users would need to add a test for the cookie in their `peek_enabled` in order to get the benefits of this PR. I see `peek_enabled` as the place to lock down which users should have access to peek.

Perhaps this is the behavior you would like, which would make this behavior opt-in and we should add this info to the README. It was my intention to turn this behavior on by default, because its zero-cognitive-load on the user and I cannot think of a case where a user would need the results when the bar is not visible.

Either way works!

mkcode commented:
`peek_enabled?` would need to look like this if we were to reuse it:

```ruby
class ApplicationController < ActionController::Base
  def peek_enabled?
    return false unless peek_bar_visible?
    current_user.staff?
  end
end
```

mkcode commented:
Hi @dewski - Hoping we can resolve this soon. What are your thoughts on https://github.com/peek/peek/pull/79#issuecomment-153131408 ?

dewski commented on app/assets/javascripts/peek.coffee:
```diff
@@ -68,4 +77,5 @@ do ($ = jQuery) ->
 
   $ ->
     if peekEnabled()
+      setPeekEnabledCookie(true)
```
We shouldn't set the `peek=true` cookie for every request as it will ignore the user's preference if they've set it to false.

dewski commented on lib/peek/controller_helpers.rb:
```diff
@@ -9,6 +9,15 @@ module ControllerHelpers
 
+    def append_info_to_payload(payload)
+      super
+      payload[:peek_bar_visible] = peek_bar_visible?
```
The only time we should be writing Peek view results into the adapter is when `peek_enabled?` returns true. Right now with the `peek_bar_visible?` guard, it still writes results for every request even if `peek_enabled?` is false.

dewski commented on lib/peek/railtie.rb:
```diff
@@ -20,7 +20,10 @@ class Railtie < ::Rails::Engine
       ActiveSupport::Notifications.subscribe('process_action.action_controller') do
-        Peek.adapter.save
+        |_, _, _, _, payload|
+        if payload[:peek_bar_visible]
```
We should update this to check if `payload[:peek_enabled]`.
